cmd/worker/main.go

```diff
@@ -92,10 +90,45 @@ func newWorker(client client.Client) worker.Worker {
 	return worker.New(client, "demo", wo)
 }
 
-func getAPIKeyFromEnv() (string, error) {
+func getAPIKeyFromEnv(ctx context.Context) (string, error) {
 	v := os.Getenv(temporalCloudAPIKeyEnvName)
 	if v == "" {
+		// if no API key is provided return an error
 		return "", fmt.Errorf("apikey not provided, set environment variable '%s' with apikey you want to use", temporalCloudAPIKeyEnvName)
 	}
 	return v, nil
 }
+
+func getNamespaceAPIKeyFromEnv(ctx context.Context) (string, error) {
+	v := os.Getenv(temporalCloudNamespaceAPIKeyEnvName)
+	if v == "" {
+		// fallback to using the control plane API key if no namespace specific API key is provided
+		v, _ = getAPIKeyFromEnv(ctx)
+	}
+	if v == "" {
+		// if no API key is provided return an error
+		return "", fmt.Errorf("namespace apikey not provided, set environment variable '%s' with apikey you want to use", temporalCloudNamespaceAPIKeyEnvName)
+	}
+	return v, nil
+}
+
+func fetchGrpcAddressUsingAPIKey(
+	ctx context.Context,
+	namespace string,
+	getAPIKeyFunc func(context.Context) (string, error),
+) (string, error) {
+	c, err := api.NewConnectionWithAPIKey(api.TemporalCloudAPIAddress, false, getAPIKeyFunc)
+	if err != nil {
+		return "", fmt.Errorf("failed to create cloud api connection: %w", err)
+	}
+	resp, err := c.CloudService().GetNamespace(ctx, &cloudservicev1.GetNamespaceRequest{
+		Namespace: namespace,
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to get namespace %q: %w", namespace, err)
+	}
+	if resp.GetNamespace().GetEndpoints().GetGrpcAddress() == "" {
```

Comment:
This is interesting that we're coupling this grpc address with the regional endpoint for api keys. Maybe we should add a comment here, since picking the right endpoint is a bit tricky.